### PR TITLE
[SL-ONLY] Feature/silabs tracing logs formating

### PR DESCRIPTION
**Pull request by lpbeliveau-silabs**
Reworked the logging, added some tests for the formatting, the error generation and the buffer handling.

Yes we are merging on main, this was a mistake I made when I made the previous one, I'll fix it tomorrow probably.

Here's an updated class diagram for all the diagram lovers out there:

```mermaid
classDiagram
    class SilabsTracer {
        +static SilabsTracer& Instance()
        +CHIP_ERROR Init()
        +CHIP_ERROR StartWatermarksStorage(PersistentStorageDelegate* storage)
        +CHIP_ERROR TimeTraceBegin(TimeTraceOperation aOperation)
        +CHIP_ERROR TimeTraceEnd(TimeTraceOperation aOperation, CHIP_ERROR error = CHIP_NO_ERROR)
        +CHIP_ERROR TimeTraceInstant(TimeTraceOperation aOperation, CHIP_ERROR error = CHIP_NO_ERROR)
        +CHIP_ERROR OutputTimeTracker(const TimeTracker& tracker)
        +CHIP_ERROR OutputWaterMark(TimeTraceOperation aOperation)
        +CHIP_ERROR OutputAllWaterMarks()
        +CHIP_ERROR TraceBufferFlushAll()
        +CHIP_ERROR TraceBufferFlushByOperation(TimeTraceOperation aOperation)
        +TimeTracker GetTimeTracker(TimeTraceOperation aOperation)
        +Watermark GetWatermark(TimeTraceOperation aOperation)
        +CHIP_ERROR SaveWatermarks()
        +CHIP_ERROR LoadWatermarks()
        +size_t GetTimeTracesCount()
        +CHIP_ERROR GetTraceByOperation(TimeTraceOperation aOperation, char * buffer, size_t & size) const
        -void TraceBufferClear()
        -CHIP_ERROR OutputTrace(const TimeTracker& tracker)
        -static SilabsTracer sInstance
        -TimeTrackerList mTimeTrackerList
        -TimeTracker mLatestTimeTrackers[kNumTraces]
        -Watermark mWatermarks[kNumTraces]
        -PersistentStorageDelegate* mStorage
        -size_t mBufferedTrackerCount
    }

    class TimeTracker {
        +System::Clock::Milliseconds32 mStartTime
        +System::Clock::Milliseconds32 mEndTime
        +TimeTraceOperation mOperation
        +OperationType mType
        +CHIP_ERROR mError
    }

    class Watermark {
        +System::Clock::Milliseconds32 mMovingAverage
        +System::Clock::Milliseconds32 mMaxTimeMs
        +System::Clock::Milliseconds32 mMinTimeMs
        +uint32_t mTotalCount
        +uint32_t mSuccessfullCount
        +uint32_t mCountAboveAvg
    }

    class TimeTrackerList {
        +chip::SingleLinkedListNode* head
        +void Insert(const TimeTracker& tracker)
        +void Remove(size_t index)
    }

    class SingleLinkedListNode {
        +TimeTracker mValue
        +SingleLinkedListNode* mpNext
    }

    SilabsTracer --> TimeTracker
    SilabsTracer --> Watermark
    SilabsTracer --> TimeTrackerList
    TimeTrackerList --> SingleLinkedListNode
    SingleLinkedListNode --> TimeTracker
```

### CODE REVIEW

File: src/platform/silabs/tracing/SilabsTracing.cpp [hunk 1]
```diff
@@ -43,6 +116,28 @@ struct TimeTrackerStorage : public TimeTracker, PersistentData<kPersistentTimeTr
     // TODO implement the Persistent Array class and use it here for logging the watermark array
 };
 
+int TimeTracker::ToCharArray(char * buffer, size_t bufferSize) const
```
Comment (mkardous-silabs) [inline src/platform/silabs/tracing/SilabsTracing.cpp:119]: I think moving this to a switch case would make it easir in the lng temps to add prints.

File: src/platform/silabs/tracing/SilabsTracing.cpp [hunk 2]
```diff
@@ -34,7 +34,80 @@ namespace chip {
 namespace Tracing {
 namespace Silabs {
 
-// TODO add string mapping to log the operation names and types
+// String mapping for TimeTraceOperation
+const char * TimeTraceOperationToString(TimeTraceOperation operation)
```
Comment (jmartinez-silabs) [inline src/platform/silabs/tracing/SilabsTracing.cpp:38]: const array could have worked too. Not sure which one is less flash hungry
```
const char* TimeTraceOperationString[TimeTraceOperation::kMaxCount] = {
    [TimeTraceOperation::kSpake2p] = "Spake2p",
    [TimeTraceOperation::kPake1]"Pake1",
    ...
};
```

File: src/platform/silabs/tracing/SilabsTracing.cpp [hunk 3]
```diff
@@ -127,28 +222,28 @@ void SilabsTracer::TimeTraceEnd(TimeTraceOperation aOperation, CHIP_ERROR error)
...
 CHIP_ERROR SilabsTracer::OutputTimeTracker(const TimeTracker & tracker)
 {
     VerifyOrReturnError(isLogInitialized(), CHIP_ERROR_UNINITIALIZED);
-    ChipLogProgress(DeviceLayer,
-                    "TimeTracker - StartTime: %" PRIu32 ", EndTime: %" PRIu32 ", Operation: %" PRIu8 ", Type: %" PRIu8
-                    ", Error: %" CHIP_ERROR_FORMAT,
-                    tracker.mStartTime.count(), tracker.mEndTime.count(), static_cast<uint8_t>(tracker.mOperation),
-                    static_cast<uint8_t>(tracker.mType), tracker.mError.Format());
+    // Allocate a buffer to store the trace
+    char buffer[256];
```
Comment (mkardous-silabs) [inline src/platform/silabs/tracing/SilabsTracing.cpp:243]: would it be useful to have a constant for the buffer size? passing a bytespan and using bytespans in the function would make sense to me

Comment (jmartinez-silabs) [inline src/platform/silabs/tracing/SilabsTracing.cpp:243]: I don't like it how we use this buffer and ToCharArray but I guess its ok

**Review by jmartinez-silabs — APPROVED** [review]

File: src/platform/silabs/tracing/SilabsTracing.cpp [hunk 4]
```diff
@@ -272,24 +369,34 @@ CHIP_ERROR SilabsTracer::LoadWatermarks()
...
+CHIP_ERROR SilabsTracer::GetTraceByOperation(TimeTraceOperation aOperation, char * buffer, size_t & size) const
 {
     auto * current = mTimeTrackerList.head;
-    for (size_t i = 0; i < index && current != nullptr; ++i)
+    while (current != nullptr)
     {
+        if (current->mValue.mOperation == aOperation)
+        {
+            // Check if the buffer is large enough
+            auto requiredSize = current->mValue.ToCharArray(nullptr, 0);
```
Comment (mkardous-silabs) [inline src/platform/silabs/tracing/SilabsTracing.cpp:380]: Having guetter that returns the size would make more sense than `ToCharArray(nullptr, 0)` which seems ominous

File: src/platform/silabs/tracing/SilabsTracing.cpp (hunk 3 above)
Comment (lpbeliveau-silabs) [reply src/platform/silabs/tracing/SilabsTracing.cpp:243]: ToCharArray and this buffer were edited to use ByteSpan and a defined max trace size.

File: src/platform/silabs/silabs_utils.cpp [hunk 5]
```diff
@@ -0,0 +1,38 @@
+/*
```
Comment (jmartinez-silabs) [inline src/platform/silabs/silabs_utils.cpp:1]: does moving this file here works with gn? No circular dependency issues? i don't see the gn change or deletion/move of the existing file.

**Review by mkardous-silabs — APPROVED** [review]

**Pull request merged**